README
This is v0.9.10 of PyECLib.  This library provides a simple Python interface for
implementing erasure codes and is known to work with Python v2.6, 2.7 and 3.x.

To obtain the best possible performance, the library utilizes liberasurecode,
which is a C based erasure code library.  Please let us know if you have any
other issues building or installing (email: kmgreen2@gmail.com or
tusharsg@gmail.com).

This library makes use of Jesasure for Reed-Solomon as implemented by the
liberasurecode library and provides its' own flat XOR-based erasure code
encoder and decoder.  Currently, it implements a specific class of HD
Combination Codes (see "Flat XOR-based erasure codes in storage systems:
Constructions, efficient recovery, and tradeoffs" in IEEE MSST 2010).  These
codes are well-suited to archival use-cases, have a simple construction and
require a minimum number of participating disks during single-disk
reconstruction (think XOR-based LRC code).

Examples of using this library are provided in "tools" directory:

  Command-line encoder::
  
      tools/pyeclib_encode.py

  Command-line decoder::
  
      tools/pyeclib_decode.py

  Utility to determine what is needed to reconstruct missing fragments::
  
      tools/pyeclib_fragments_needed.py


PyEClib initialization::

  ec_driver = ECDriver(k=<num_encoded_data_fragments>,
                       m=<num_encoded_parity_fragments>,
                       ec_type=<ec_scheme>))

Supported ``ec_type`` values:

  * ``jerasure_rs_vand`` => Vandermonde Reed-Solomon encoding
  * ``flat_xor_hd_3``, ``flat_xor_hd_4`` => Flat-XOR based HD combination codes

A configuration utility is provided to help compare available EC schemes in 
terms of performance and redundancy:: tools/pyeclib_conf_tool.py


The Python API supports the following functions:

- EC Encode

  Encode N bytes of a data object into k (data) + m (parity) fragments::

    def encode(self, data_bytes)

    input:   data_bytes - input data object (bytes)
    returns: list of fragments (bytes)


- EC Decode

  Decode between k and k+m fragments into original object::

    def decode(self, fragment_payloads)

    input:   list of fragment_payloads (bytes)
    returns: decoded object (bytes)


*Note*: ``bytes`` is a synonym to ``str`` in Python 2.6, 2.7.
In Python 3.x, ``bytes`` and ``str`` types are non-interchangeable and care
needs to be taken when handling input to and output from the ``encode()`` and
``decode()`` routines.


- EC Reconstruct

  Reconstruct "missing_fragment_indexes" using "available_fragment_payloads"::

    def reconstruct(self, available_fragment_payloads, missing_fragment_indexes)


- Minimum parity fragments needed for durability gurantees
    
    def min_parity_fragments_needed(self)


- Fragments needed for EC Reconstruct

  Return the indexes of fragments needed to reconstruct "missing_fragment_indexes"::

    def fragments_needed(self, missing_fragment_indexes)


- Get EC Metadata

  Return an opaque buffer known by the underlying library::

    def get_metadata(self, fragment)


- Verify EC Stripe Consistency

  Use opaque buffers from get_metadata() to verify a the consistency of a stripe::

    def verify_stripe_metadata(self, fragment_metadata_list)


- Get EC Segment Info

  Return a dict with the keys - segment_size, last_segment_size, fragment_size, last_fragment_size and num_segments::

    def get_segment_info(self, data_len, segment_size)


Quick Start:

  Standard stuff to install::
  
    ``Python 2.6``, ``2.7`` or ``3.x`` (including development packages), ``argparse`` and ``liberasurecode``.


  As mentioned above, PyECLib depends on the installation of the liberasurecde library (liberasurecode
  can be found at http://bitbucket.org/elambert/liberasurecode).


  Install PyECLib::

    $ sudo python setup.py install

  Run test suite included::

    $ sudo python setup.py test && (cd test; ./ec_pyeclib_file_test.sh)

  If all of this works, then you should be good to go.  If not, send us an email!

  If the test suite fails because it cannot find any of the shared libraries,
  then you probably need to add /usr/local/lib to the path searched when loading
  libraries.  The best way to do this (on Linux) is to add '/usr/local/lib' to::

    /etc/ld.so.conf 

  and then run::

    $ ldconfig

--
0.10


debian/changelog
python-pyeclib (1.4.0-2~bpo8+1) jessie-backports; urgency=medium

  * Rebuild for jessie-backports.

 -- Ondřej Nový <onovy@debian.org>  Tue, 21 Feb 2017 12:40:12 +0100

python-pyeclib (1.4.0-2) unstable; urgency=medium

  * Prevent division by zero in tests (patch from upstream)

 -- Ondřej Nový <onovy@debian.org>  Sun, 25 Dec 2016 21:01:35 +0100

python-pyeclib (1.4.0-1) unstable; urgency=medium

  * New upstream release
  * Bump debhelper version to 10
  * debian/tests/control: Add isa_l_rs_cauchy backend test

 -- Ondřej Nový <onovy@debian.org>  Tue, 13 Dec 2016 23:30:09 +0100

python-pyeclib (1.3.1-1) unstable; urgency=medium

  * New upstream release
  * Removed Pre-Depends on dpkg

 -- Ondřej Nový <onovy@debian.org>  Sat, 08 Oct 2016 20:28:22 +0200

python-pyeclib (1.2.1-2) unstable; urgency=medium

  * d/{control,copyright}: Use my @debian.org email address
  * Fixed .gitreview file
  * d/s/options: extend-diff-ignore of .gitreview
  * d/control: Using OpenStack's Gerrit as VCS URLs.

 -- Ondřej Nový <onovy@debian.org>  Fri, 23 Sep 2016 21:03:05 +0200

python-pyeclib (1.2.1-1) unstable; urgency=medium

  * New upstream release
  * Upstream moved to official Openstack project
  * Bump liberasurecode-dev dependency to 1.2.0 version
  * Added Flat XOR HD backend to Debian test
  * Print debug info in Debian test

 -- Ondřej Nový <novy@ondrej.org>  Thu, 02 Jun 2016 12:49:11 +0200

python-pyeclib (1.2.0-2) unstable; urgency=medium

  * Tests liberasure backends in Debian tests.
  * Removed Suggests libjerasure2.
  * Standards-Version is 3.9.8 now (no change).

 -- Ondřej Nový <novy@ondrej.org>  Fri, 08 Apr 2016 23:38:18 +0200

python-pyeclib (1.2.0-1) unstable; urgency=medium

  * New upstream release.
  * Added Suggests libjerasure2.
  * Added Debian tests.
  * Added myself as uploader.
  * Added myself to copyright file.
  * Fixed Vcs URLs.
  * Removed patch: disable-check-for-liberasure-code.patch
  * Removed patch: more-rpath-tweak-removal.patch
  * Removed install-liberasurecode=no option, it's not needed anymore.
  * Fixed d/copyright.
  * Standards-Version is 3.9.7 now (no change).

 -- Ondřej Nový <novy@ondrej.org>  Fri, 26 Feb 2016 22:29:07 +0100

python-pyeclib (1.1.1-2) unstable; urgency=medium

  * Rebuild with liberasurecode 1.1.0.

 -- Thomas Goirand <zigo@debian.org>  Fri, 11 Dec 2015 23:54:00 +0000

python-pyeclib (1.1.1-1) unstable; urgency=medium

  * New upstream release.
  * Add --install-liberasurecode=no when installing with "python setup.py
    install".
  * Drop version of python-all-dev build-depends.
  * Dropped big-endian.patch applied upstream.
  * Rebased patches:
    - disable-check-for-liberasure-code.patch
    - more-rpath-tweak-removal.patch
  * Overrides dh_auto_install, not dh_install.

 -- Thomas Goirand <zigo@debian.org>  Tue, 10 Nov 2015 11:08:06 +0000

python-pyeclib (1.0.8-3) unstable; urgency=medium

  * Team upload.
  * Drop BD's on gf-complete and jerasure dev packages; not required
    due to the runtime nature of backend loading in liberasurecode.
  * d/p/big-endian.patch: Cherry pick fix from upstream VCS to resolve
    issues on big endian architectures.

 -- James Page <james.page@ubuntu.com>  Wed, 09 Sep 2015 13:32:08 +0100

python-pyeclib (1.0.8-2) unstable; urgency=medium

  * Team upload.
  * Match version of liberasurecode-dev with pyeclib.

 -- James Page <james.page@ubuntu.com>  Thu, 06 Aug 2015 14:19:02 +0100

python-pyeclib (1.0.8-1) unstable; urgency=medium

  [ James Page ]
  * d/rules: Enable unit testing for python2.

  [ Thomas Goirand ]
  * New upstream release.
  * Rebased patches.

 -- Thomas Goirand <zigo@debian.org>  Wed, 05 Aug 2015 08:35:22 +0000

python-pyeclib (1.0.7m-1) unstable; urgency=medium

  * New upstream release.
  * Added dh-python as build-depends.

 -- Thomas Goirand <zigo@debian.org>  Sun, 17 May 2015 05:55:24 +0200

python-pyeclib (1.0.5m-1) unstable; urgency=medium

  * New upstream release using liberasurecode 1.0.5.
  * Refreshed patches.

 -- Thomas Goirand <zigo@debian.org>  Fri, 13 Mar 2015 23:23:29 +0000

python-pyeclib (0.9.10-3) unstable; urgency=medium

  * Fixed the debian/watch file to use the URL from bitbucket.

 -- Thomas Goirand <zigo@debian.org>  Sun, 08 Mar 2015 18:41:06 +0000

python-pyeclib (0.9.10-2) unstable; urgency=medium

  * Added missing python{3,}-all-dev build-depends.
  * Added missing libjerasure-dev build-depends.

 -- Thomas Goirand <zigo@debian.org>  Thu, 20 Nov 2014 02:15:48 +0800

python-pyeclib (0.9.10-1) unstable; urgency=low

  * Initial release. (Closes: #769519)

 -- Thomas Goirand <zigo@debian.org>  Fri, 14 Nov 2014 17:37:50 +0800


test/test_core.py
# Copyright (c) 2013, Kevin Greenan (kmgreen2@gmail.com)
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.  THIS SOFTWARE IS
# PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
# OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
# NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from functools import wraps
import os
import sys
import unittest

run_under_valgrind = False
test_cmd_prefix = ""
log_filename_prefix = ""


class TestCore(unittest.TestCase):

    def __init__(self, *args):
        self.pyeclib_core_test = "test_pyeclib_c.py"
        self.pyeclib_iface_test = "test_pyeclib_api.py"

        unittest.TestCase.__init__(self, *args)

    def setUp(self):
        # Determine which directory we're in
        dirs = os.getcwd().split('/')
        if dirs[-1] == 'test':
            self.pyeclib_test_dir = "."
        else:
            self.pyeclib_test_dir = "./test"

        # Create the array of tests to run
        self.py_test_dirs = [
            (self.pyeclib_test_dir, self.pyeclib_core_test),
            (self.pyeclib_test_dir, self.pyeclib_iface_test)
        ]

    def tearDown(self):
        pass

    def test_core(self):
        self.assertTrue(True)
        cur_dir = os.getcwd()
        print("\n")
        for (dir, test) in self.py_test_dirs:
            sys.stdout.write("Running test %s ... " % test)
            sys.stdout.flush()
            os.chdir(dir)
            if os.path.isfile(test):
                ret = os.system(
                    "%s python %s >%s/%s.%s.out 2>&1" %
                    (test_cmd_prefix, test, cur_dir,
                     log_filename_prefix, test))

                self.assertEqual(0, ret)
                os.system("rm -f *.pyc")
                os.chdir(cur_dir)
                print('ok')
            else:
                self.assertTrue(False)
                print('failed')


# Invoke this script as "python test_core_valgrind.py"
# for the "valgrind" variant
# (test_core_valgrind.py is a symlink to test_core.py)
if __name__ == "__main__":
    if '_valgrind' in sys.argv[0]:
        if (0 != os.system("which valgrind")):
            print("You don't appear to have 'valgrind' installed")
            sys.exit(-1)
        run_under_valgrind = True
        test_cmd_prefix = "valgrind --leak-check=full "
        log_filename_prefix = "valgrind"
    if sys.version_info<(2,7,0):
        unittest.main()
    else:
        unittest.main(verbosity=2)


debian/tests/encode-decode.py
from __future__ import print_function
from pyeclib.ec_iface import ECDriver
import sys

input = b'test'

# Init
print("init:", end=" ")
ec = ECDriver(k=3, m=3, hd=3, ec_type=sys.argv[1])
print("OK")

# Encode
print("encode:", end=" ")
fragments = ec.encode(input)
print("OK")

# Decode
print("decode:", end=" ")
assert ec.decode(fragments[0:ec.k]) == input
print("OK")


tools/pyeclib_conf_tool.py
# Copyright (c) 2013, Kevin Greenan (kmgreen2@gmail.com)
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.  THIS SOFTWARE IS
# PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
# OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
# NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


# PyEClib Companion tool
# Goal: When defining an EC pool, help cluster admin make an informed choice
# between available EC implementations. Generate sample swift.conf + swift-
# ring-builder hints.
#
# Suggested features:
#
# - List the "EC types" supported - EC algorithms
# - List implementations of each EC type available on the platform
#   (dumb-software-only, software with SIMD acceleration,
#    specialized hardware, etc).
# - Benchmark each algorithm with possible implementation and display
#   performance numbers.
# - Generate sample EC policy entry (for inclusion in swift.conf) for the
#   best performing algorithm + implementation. (And optionally provide swift-
#   ring-builder hints).
#
# Suggested EC policy entry format:
#
# ======== swift.conf ============
# [storage-policy:10]
# type = erasure_coding
# name = ec_jerasure_rs_cauchy_12_2
# ec_type = jerasure_rs_cauchy
# ec_k = 12
# ec_m = 2
# ============================
#
# (ec_type values are one of those available within PyEClib)

#
# User input: Num data, num parity, average file size
# Output: Ordered list of options and their corresponding conf entries
#         (limit 10)
#

import pyeclib
from pyeclib.ec_iface import ECDriver
import random
import string
import sys
import os
import argparse
import time
import math


class Timer:

    def __init__(self):
        self.start_time = 0
        self.end_time = 0

    def reset(self):
        self.start_time = 0
        self.end_time = 0

    def start(self):
        self.start_time = time.time()

    def stop(self):
        self.end_time = time.time()

    def curr_delta(self):
        return self.end_time - self.start_time

    def stop_and_return(self):
        self.end_time = time.time()
        return self.curr_delta()


def nCr(n, r):
    f = math.factorial
    return f(n) / f(r) / f(n - r)


class ECScheme:

    def __init__(self, k, m, w, ec_type):
        self.k = k
        self.m = m
        self.w = w
        self.ec_type = ec_type

    def __str__(self):
        return "k=%d m=%d w=%d ec_type=%s" % (self.k, self.m, self.w, self.ec_type)

valid_flat_xor_hd_3 = [(6, 6), (7, 6), (8, 6), (9, 6),
                    (10, 6), (11, 6), (12, 6), (13, 6),
                    (14, 6), (15, 6)]

valid_flat_xor_hd_4 = [(6, 6), (7, 6), (8, 6), (9, 6),
                    (10, 6), (11, 6), (12, 6), (13, 6),
                    (14, 6), (15, 6), (16, 6), (17, 6),
                    (18, 6), (19, 6), (20, 6)]


def get_viable_schemes(
        max_num_frags, minimum_rate, avg_stripe_size, fault_tolerance):

    list_of_schemes = []

    #
    # Get min_k from (minimum_rate * max_num_frags)
    #
    min_k = int(math.ceil(minimum_rate * max_num_frags))

    #
    # Get min_m from the fault tolerance
    #
    min_m = fault_tolerance

    #
    # Is not information theoretically possible
    #
    if (min_k + min_m) > max_num_frags:
        return list_of_schemes

    #
    # Iterate over EC(k, max_num_frags-k) k \in [min_k, n-min_m]
    #
    for k in range(min_k, max_num_frags - min_m + 1):
        #
        # RS(k, max_num_frags-k) is trivial, just add it
        # (w=[8,16,32] for vand_rs)
        #
        for w in [8, 16, 32]:
            list_of_schemes.append(
                ECScheme(k, max_num_frags - k, w, "jerasure_rs_vand_%d" % w))

        for w in [4, 8]:
            list_of_schemes.append(
                ECScheme(k, max_num_frags - k, w, "jerasure_rs_cauchy_%d" % w))

        #
        # The XOR codes are a little tricker
        # (only check if fault_tolerance = 2 or 3)
        #
        # Constraint for 2: k <= (m choose 2)
        # Constraint for 3: k <= (m choose 3)
        #
        # The '3' flat_xor_hd_3  (and '4' in flat_xor_hd_4) refers to the Hamming
        # distance, which means the code guarantees the reconstruction of any
        # 2 lost fragments (or 3 in the case of flat_xor_hd_4).
        #
        # So, only consider the XOR code if the fault_tolerance matches and
        # the additional constraint is met
        #
        if fault_tolerance == 2:
            max_k = nCr(max_num_frags - k, 2)
            if k <= max_k and (k, max_num_frags - k) in valid_flat_xor_hd_3:
                list_of_schemes.append(
                    ECScheme(k, max_num_frags - k, 0, "flat_xor_hd_3"))

        if fault_tolerance == 3:
            max_k = nCr(max_num_frags - k, 3)
            if k <= max_k and (k, max_num_frags - k) in valid_flat_xor_hd_4:
                list_of_schemes.append(
                    ECScheme(k, max_num_frags - k, 0, "flat_xor_hd_4"))

    return list_of_schemes


parser = argparse.ArgumentParser(
    description='PyECLib tool to evaluate viable EC options, benchmark them '
                'and report results with the appropriate conf entries.')
parser.add_argument(
    '-n',
    type=int,
    help='max number of fragments',
    required=True)
parser.add_argument('-f', type=int, help='fault tolerance', required=True)
parser.add_argument(
    '-r',
    type=float,
    help='minimum coding rate (num_data / num_data+num_parity)',
    required=True)
parser.add_argument('-s', type=int, help='average stripe size', required=True)
parser.add_argument(
    '-l',
    type=int,
    help='set limit on number of entries returned (default = 10)',
    default=10,
)

args = parser.parse_args(sys.argv[1:])

MB = 1024 * 1024

# Generate a buffer of size 's'
if args.s > 10 * MB:
    print("s must be smaller than 10 MB.")
    sys.exit(1)

# Instantiate the timer
timer = Timer()

return_limit = args.l

schemes = get_viable_schemes(args.n, args.r, args.s, args.f)

# Results will be List[(ec_type, throughput)]
results = []

# Num iterations
num_iterations = 10

for scheme in schemes:
    print(scheme)

    # Generate a new string for each test
    file_str = ''.join(
        random.choice(
            string.ascii_uppercase + string.digits) for x in range(args.s))

    try:
        ec_driver = ECDriver(k=scheme.k, m=scheme.m, ec_type=scheme.ec_type)
    except Exception as e:
        print("Scheme %s is not defined (%s)." % (scheme, e))
        continue

    timer.start()

    for i in range(num_iterations):
        ec_driver.encode(file_str)

    duration = timer.stop_and_return()

    results.append((scheme, duration))

    timer.reset()

results.sort(key=lambda x, y: (int)((1000 * x[1]) - (1000 * y[1])))

for i in range(len(results)):
    if i > return_limit:
        break

    print("\n\nPerf Rank #%d:" % i)
    print("  ======== To Use this Policy, Copy and Paste Text (not including "
          "this header and footer) to Swift Conf ========")
    print("  type = erasure_coding")
    print("  name = %s_%d_%d" % (results[i][0].ec_type,
                                 results[i][0].k, results[i][0].m))
    print("  ec_type = %s" % results[i][0].ec_type)
    print("  ec_k = %s" % results[i][0].k)
    print("  ec_m = %s" % results[i][0].m)
    print("  ================================================================"
          "==============================================")
    results[i]


src/python/pyeclib/ec_iface.py
# Copyright (c) 2013-2014, Kevin Greenan (kmgreen2@gmail.com)
# Copyright (c) 2014, Tushar Gohad (tushar.gohad@intel.com)
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.  THIS SOFTWARE IS
# PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
# OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
# NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from enum import Enum
from enum import unique
from utils import create_instance
from utils import positive_int_value


def PyECLibVersion(z, y, x):
    return (((z) << 16) + ((y) << 8) + (x))

PYECLIB_MAJOR = 0
PYECLIB_MINOR = 9
PYECLIB_REV = 4
PYECLIB_VERSION = PyECLibVersion(PYECLIB_MAJOR, PYECLIB_MINOR,
                                 PYECLIB_REV)


PYECLIB_MAX_DATA = 32
PYECLIB_MAX_PARITY = 32


@unique
class PyECLibEnum(Enum):

    def describe(self):
        # returns supported types
        return list(self)

    @classmethod
    def has_enum(cls, name):
        # returns True if name is a valid member of the enum
        try:
            cls.__getattr__(name)
        except AttributeError:
            return False
        return True

    @classmethod
    def get_by_name(cls, name):
        try:
            obj = cls.__getattr__(name)
        except AttributeError:
            return None
        return obj

    @classmethod
    def names(cls):
        return [name for name, value in cls.__members__.items()]

    @classmethod
    def values(cls):
        return [value for name, value in cls.__members__.items()]

    def __str__(self):
        return "%s: %d" % (self.name, self.value)


# Erasure Code backends supported as of this PyECLib API rev
class PyECLib_EC_Types(PyECLibEnum):
    # Note: the Enum start value defaults to 1 as the starting value and not 0
    # 0 is False in the boolean sense but enum members evaluate to True
    jerasure_rs_vand = 1
    jerasure_rs_cauchy = 2
    flat_xor_hd = 3
    isa_l_rs_vand = 4


# Output of Erasure (en)Coding process are data "fragments".  Fragment data
# integrity checks are provided by a checksum embedded in a header (prepend)
# for each fragment.

# The following Enum defines the schemes supported for fragment checksums.
# The checksum type is "none" unless specified.
class PyECLib_FRAGHDRCHKSUM_Types(PyECLibEnum):
    # Note: the Enum start value defaults to 1 as the starting value and not 0
    # 0 is False in the boolean sense but enum members evaluate to True
    none = 1
    algsig = 2
    inline_crc32 = 3


# Generic ECDriverException
class ECDriverError(Exception):

    def __init__(self, error_str):
        self.error_str = error_str

    def __str__(self):
        return self.error_str


# Main ECDriver class
class ECDriver(object):

    def __init__(self, *args, **kwargs):
        self.k = -1
        self.m = -1
        self.ec_type = None
        self.chksum_type = None
        self.library_import_str = None
        for (key, value) in kwargs.items():
            if key == "k":
                try:
                    self.k = positive_int_value(value)
                except ValueError as e:
                    raise ECDriverError(
                        "Invalid number of data fragments (k)")
            elif key == "m":
                try:
                    self.m = positive_int_value(value)
                except ValueError as e:
                    raise ECDriverError(
                        "Invalid number of data fragments (m)")
            elif key == "ec_type":
                if PyECLib_EC_Types.has_enum(value):
                    self.ec_type = \
                        PyECLib_EC_Types.get_by_name(value)
                else:
                    raise ECDriverError(
                        "%s is not a valid EC type for PyECLib!" % value)
            elif key == "chksum_type":
                if PyECLib_FRAGHDRCHKSUM_Types.has_enum(value):
                    self.chksum_type = \
                        PyECLib_FRAGHDRCHKSUM_Types.get_by_name(value)
                else:
                    raise ECDriverError(
                        "%s is not a valid checksum type for PyECLib!" % value)

        if library_import_str is not None:
            self.library_import_str = library_import_str
        else:
            raise ECDriverError(
                "Library import string (library_import_str) was not specified "
                "and is a required argument!")
        #
        # Instantiate EC backend driver
        #
        self.ec_lib_reference = create_instance(
            library_import_str,
            k=self.k,
            m=self.m,
            ec_type=self.ec_type,
            chksum_type=self.chksum_type)
        #
        # Verify that the imported library implements the required functions
        #
        required_methods = {
            'decode': 0,
            'encode': 0,
            'reconstruct': 0,
            'fragments_needed': 0,
            'min_parity_fragments_needed': 0,
            'get_metadata': 0,
            'verify_stripe_metadata': 0,
            'get_segment_info': 0
        }

        for attr in dir(self.ec_lib_reference):
            if hasattr(getattr(self.ec_lib_reference, attr), "__call__"):
                required_methods[attr] = 1

        not_implemented_str = ""
        for (method, is_implemented) in required_methods.items():
            if is_implemented == 0:
                not_implemented_str += method + " "

        if len(not_implemented_str) > 0:
            raise ECDriverError(
                "The following required methods are not implemented "
                "in %s: %s" % (self.library_import_str, not_implemented_str))

    def encode(self, data_bytes):
        """
        Encode an arbitrary-sized string
        :param data_bytes: the buffer to encode
        :returns: a list of buffers (first k entries are data and
                  the last m are parity)
        :raises: ECDriverError if there is an error during encoding
        """
        return self.ec_lib_reference.encode(data_bytes)

    def decode(self, fragment_payloads):
        """
        Decode a set of fragments into a buffer that represents the original
        buffer passed into encode().

        :param fragment_payloads: a list of buffers representing a subset of
                                  the list generated by encode()
        :returns: a buffer
        :raises: ECDriverError if there is an error during decoding
        """
        return self.ec_lib_reference.decode(fragment_payloads)

    def reconstruct(self, available_fragment_payloads,
                    missing_fragment_indexes):
        """
        Reconstruct a missing fragment from a subset of available fragments.

        :param available_fragment_payloads: a list of buffers representing
                                            a subset of the list generated
                                            by encode()
        :param missing_fragment_indexes: a list of integers representing
                                         the indexes of the fragments to be
                                         reconstructed.
        :param destination_index: the index of the element to reconstruct
        :returns: a list of buffers (ordered by fragment index) containing
                  the reconstructed payload associated with the indexes
                  provided in missing_fragment_indexes
        :raises: ECDriverError if there is an error during decoding or there
                 are not sufficient fragments to decode
        """
        return self.ec_lib_reference.reconstruct(
            available_fragment_payloads, missing_fragment_indexes)

    def fragments_needed(self, reconstruction_indexes, exclude_indexes = []):
        """
        Determine which fragments are needed to reconstruct some subset of
        missing fragments.

        :param reconstruction_indexes: a list of integers representing the
                                         indexes of the fragments to be
                                         reconstructed.
        :param exclude_indexes: a list of integers representing the
                                         indexes of the fragments to be
                                         excluded from the reconstruction
                                         equations. 
        :returns: a list containing fragment indexes that can be used to 
                  reconstruct the missing fragments.
        :raises: ECDriverError if there is an error during decoding or there
                 are not sufficient fragments to decode
        """
        return self.ec_lib_reference.fragments_needed(reconstruction_indexes, exclude_indexes)

    def min_parity_fragments_needed(self):
        return self.ec_lib_reference.min_parity_fragments_needed()

    def get_metadata(self, fragment):
        """
        Get opaque metadata for a fragment.  The metadata is opaque to the
        client, but meaningful to the underlying library.  It is used to verify
        stripes in verify_stripe_metadata().

        :param fragment: a buffer representing a single fragment generated by
                         the encode() function.
        :returns: an opaque buffer to be passed into verify_stripe_metadata()
        :raises: ECDriverError if there was a problem getting the metadata.
        """
        return self.ec_lib_reference.get_metadata(fragment)

    def verify_stripe_metadata(self, fragment_metadata_list):
        """
        Verify a subset of fragments generated by encode()

        :param fragment_metadata_list: a list of buffers representing the
                                       metadata from a subset of the framgments
                                       generated by encode().
        :returns: 'None' if the metadata is consistent.
                  a list of fragment indexes corresponding to inconsistent
                  fragments
        :raises: ECDriverError if there was a problem verifying the metadata

        """
        return self.ec_lib_reference.verify_stripe_metadata(
            fragment_metadata_list)

    def get_segment_info(self, data_len, segment_size):
        """
        Get segmentation info for a given data length and
        segment size.

        Semment info returns a dict with the following keys:

        segment_size: size of the payload to give to encode()
        last_segment_size: size of the payload to give to encode()
        fragment_size: the fragment size returned by encode()
        last_fragment_size: the fragment size returned by encode()
        num_segments: number of segments

        This allows the caller to prepare requests
        when segmenting a data stream to be EC'd.

        Since the data length will rarely be aligned
        to the segment size, the last segment will be
        a different size than the others.

        There are restrictions on the length given to encode(),
        so calling this before encode is highly recommended when
        segmenting a data stream.
        """
        return self.ec_lib_reference.get_segment_info(data_len, segment_size)

    #
    # Map of segment indexes with a list of tuples
    #
    def get_segment_info_byterange(self, ranges, data_len, segment_size):
        """
        Get segmentation info for a byterange request, given a data length and
        segment size.

        This will return a map-of-maps that represents a recipe describing 
        the segments and ranges within each segment needed to satisfy a range
        request.

        Assume a range request is given for an object with segment size 3K and
        a 1 MB file:

        Ranges = (0, 1), (1, 12), (10, 1000), (0, segment_size-1), 
                 (1, segment_size+1), (segment_size-1, 2*segment_size)

        This will return a map keyed on the ranges, where there is a recipe
        given for each range:

        {
         (0, 1): {0: (0, 1)}, 
         (10, 1000): {0: (10, 1000)}, 
         (1, 12): {0: (1, 12)}, 
         (0, 3071): {0: (0, 3071)}, 
         (3071, 6144): {0: (3071, 3071), 1: (0, 3071), 2: (0, 0)}, 
         (1, 3073): {0: (1, 3071), 1: (0,0)}
        }

        """
        
        segment_info = self.ec_lib_reference.get_segment_info(data_len, segment_size)

        segment_size = segment_info['segment_size']
        last_segment_size = segment_info['last_segment_size']
        fragment_size = segment_info['fragment_size']
        last_fragment_size = segment_info['last_fragment_size']
        num_segments = segment_info['num_segments']

        sorted_ranges = ranges[:]
        sorted_ranges.sort(lambda x, y: x[0] - y[0])

        recipe = {}

        for r in ranges:
            segment_map = {}
            begin_off = r[0]
            end_off = r[1]
            begin_segment = begin_off / segment_size 
            end_segment = end_off / segment_size 
          
            if begin_segment == end_segment:
                begin_relative_off = begin_off % segment_size
                end_relative_off = end_off % segment_size
                segment_map[begin_segment] = (begin_relative_off, end_relative_off)
            else:
                begin_relative_off = begin_off % segment_size
                end_relative_off = end_off % segment_size
                
                segment_map[begin_segment] = (begin_relative_off, segment_size-1)

                for middle_segment in range(begin_segment+1, end_segment):
                    segment_map[middle_segment] = (0, segment_size-1)
                
                segment_map[end_segment] = (0, end_relative_off)

            recipe[r] = segment_map
               

        return recipe


